src/core/rbtree.rs

```diff
@@ -0,0 +1,358 @@
+//! Wrapper over the `ngx_rbtree_t`.
+
+use core::alloc::Layout;
+use core::cmp::Ordering;
+use core::hash::{self, BuildHasher, Hash};
+use core::marker::PhantomData;
+use core::ptr::{self, NonNull};
+use core::{borrow, mem};
+
+use nginx_sys::{
+    ngx_rbt_red, ngx_rbtree_data, ngx_rbtree_delete, ngx_rbtree_init, ngx_rbtree_insert,
+    ngx_rbtree_key_t, ngx_rbtree_min, ngx_rbtree_next, ngx_rbtree_node_t, ngx_rbtree_t,
+};
+
+use crate::allocator::{self, AllocError, Allocator};
+
+#[allow(deprecated)]
+type RbTreeHasher = core::hash::BuildHasherDefault<hash::SipHasher>;
+
+/// Wrapper over the `ngx_rbtree_t`.
+#[derive(Debug)]
+pub struct RbTree<K, V, A>
+where
+    A: Allocator,
+{
+    tree: ngx_rbtree_t,
+    sentinel: NonNull<ngx_rbtree_node_t>,
+    alloc: A,
+    _ph: PhantomData<(K, V)>,
```

Comment:
I'd probably name this field _kv_type to make it clear what the purpose of the PhantomData is